common/src/main/java/com/bakdata/quick/common/api/client/PartitionedMirrorClient.java

```
@@ -0,0 +1,183 @@
+                    host.forAll(), this.parser::deserializeList), Collections.emptyList());
+            valuesFromAllHosts.addAll(valuesFromSingleHost);
+        }
+        return valuesFromAllHosts;
+    }
+
+    @Override
+    @Nullable
+    public List<V> fetchValues(final List<K> keys) {
+        return keys.stream().map(this::fetchValue).collect(Collectors.toList());
+    }
+
+    @Override
+    public boolean exists(final K key) {
+        return this.fetchValue(key) != null;
+    }
+
+    @Nullable
+    @Override
+    public <T> T sendRequest(final String url, final ParserFunction<T> parser) {
+        return this.delegate.sendRequest(url, parser);
+    }
+
+    @Nullable
+    @Override
+    public ResponseBody makeRequest(final String url) {
+        return this.delegate.makeRequest(url);
+    }
+
+    private void initRouter() {
+        log.info("Initializing partition router...");
+        final Map<Integer, String> response = makeRequestForPartitionHostMapping();
+        this.router = new PartitionRouter<>(this.keySerde, this.topicName, this.partitionFinder, response);
+        this.knownHosts = this.router.getAllHosts();
+    }
+
+    private Map<Integer, String> makeRequestForPartitionHostMapping() {
+        final String url = this.streamsStateHost.getPartitionToHostUrl();
+        try (final ResponseBody responseBody = makeRequest(url)) {
+            final TypeReference<Map<Integer, String>> typeRef = new TypeReference<>() {};
+            final Map<Integer, String> partitionHostMappingResponse = this.client.objectMapper().readValue(
+                    Objects.requireNonNull(responseBody).byteStream(), typeRef);
+            log.info("Collected information about the partitions and hosts."
+                            + " There are {} partitions and {} distinct hosts", partitionHostMappingResponse.size(),
+                    (int) partitionHostMappingResponse.values().stream().distinct().count());
```

This is potentially somewhat costly. You could avoid unnecessary calculation by wrapping it in an if statement

---

Nope:
```
final String url = this.streamsStateHost.getPartitionToHostUrl();
        Map<Integer, String> partitionHostMappingResponse = new HashMap<>();
        try (final ResponseBody responseBody = requestManager.makeRequest(url)) {
            if (responseBody != null) {
                final TypeReference<Map<Integer, String>> typeRef = new TypeReference<>() {};
                partitionHostMappingResponse = this.client.objectMapper()
                    .readValue(Objects.requireNonNull(responseBody).byteStream(), typeRef);
                log.info("Collected information about the partitions and hosts."
                        + " There are {} partitions and {} distinct hosts", partitionHostMappingResponse.size(),
                    (int) partitionHostMappingResponse.values().stream().distinct().count());
            } else {
                log.info("Response body was null. Returning an empty mapping.");
            }
            return partitionHostMappingResponse;
        } catch (final IOException e) {
            throw new InternalErrorException("There was a problem handling the response: " + e.getMessage());
        }
```

---

But probably the empty mapping is senseless, so I think I would rather throw an exception there.

---

```
        try (final ResponseBody responseBody = requestManager.makeRequest(url)) {
            if (responseBody != null) {
                final TypeReference<Map<Integer, String>> typeRef = new TypeReference<>() {};
                final Map<Integer, String> partitionHostMappingResponse = this.client.objectMapper()
                    .readValue(Objects.requireNonNull(responseBody).byteStream(), typeRef);
                log.info("Collected information about the partitions and hosts."
                        + " There are {} partitions and {} distinct hosts", partitionHostMappingResponse.size(),
                    (int) partitionHostMappingResponse.values().stream().distinct().count());
                return partitionHostMappingResponse;
            } else {
                throw new InternalErrorException("Response body was null. Returning an empty mapping.");
                
            }
```

---

I was referring to ` (int) partitionHostMappingResponse.values().stream().distinct().count()` in the logging statement

---

Ok, I thought you were referring to the fact that makeRequest can return null - which I didn't cover btw.
Basically we can throw the info about the number of hosts away if you think it is costly. On the other hand, the number of replicas won't change dynamically (will it?), so I guess you won't compute this very often

---

You can wrap it in  `if(log.isInfoEnabled) {...}`.

For the null case: Good catch, though `/partitions` should never return `null`. I'd also suggest exiting early, e.g.:
```java
if (responseBody == null) {
  throw new InternalErrorException(...)
}

// here conversion to map
```